File: tests/rosetta/transpiler/Java/cholesky-decomposition-1.java
public class Main {

    static double sqrtApprox(double x) {
        double guess = x;
        int i = 0;
        while (i < 20) {
            guess = (guess + x / guess) / 2.0;
            i = i + 1;
        }
        return guess;
    }

    static java.util.Map<String,Object> makeSym(int order, double[] elements) {
        return new java.util.LinkedHashMap<String, Object>(java.util.Map.ofEntries(java.util.Map.entry("order", order), java.util.Map.entry("ele", elements)));
    }

    static double[][] unpackSym(java.util.Map<String,Object> m) {
        Object n = (Object)(((Object)m.get("order")));
        Object ele = (Object)(((Object)m.get("ele")));
        double[][] mat = new double[][]{};
        int idx = 0;
        int r = 0;
        while (r < ((Number)(n)).intValue()) {
            double[] row = new double[]{};
            int c = 0;
            while (c <= r) {
                row = java.util.stream.DoubleStream.concat(java.util.Arrays.stream(row), java.util.stream.DoubleStream.of(((Number)(ele[idx])).doubleValue())).toArray();
                idx = idx + 1;
                c = c + 1;
            }
            while (c < ((Number)(n)).intValue()) {
                row = java.util.stream.DoubleStream.concat(java.util.Arrays.stream(row), java.util.stream.DoubleStream.of(0.0)).toArray();
                c = c + 1;
            }
            mat = appendObj(mat, row);
            r = r + 1;
        }
        r = 0;
        while (r < ((Number)(n)).intValue()) {
            int c = r + 1;
            while (c < ((Number)(n)).intValue()) {
mat[r][c] = mat[c][r];
                c = c + 1;
            }
            r = r + 1;
        }
        return mat;
    }

    static void printMat(double[][] m) {
        int i = 0;
        while (i < m.length) {
            String line = "";
            int j = 0;
            while (j < m[i].length) {
                line = line + String.valueOf(m[i][j]);
                if (j < m[i].length - 1) {
                    line = line + " ";
                }
                j = j + 1;
            }
            System.out.println(line);
            i = i + 1;
        }
    }

    static void printSym(java.util.Map<String,Object> m) {
        printMat(unpackSym(m));
    }

    static void printLower(java.util.Map<String,Object> m) {
        Object n = (Object)(((Object)m.get("order")));
        Object ele = (Object)(((Object)m.get("ele")));
        double[][] mat = new double[][]{};
        int idx = 0;
        int r = 0;
        while (r < ((Number)(n)).intValue()) {
            double[] row = new double[]{};
            int c = 0;
            while (c <= r) {
                row = java.util.stream.DoubleStream.concat(java.util.Arrays.stream(row), java.util.stream.DoubleStream.of(((Number)(ele[idx])).doubleValue())).toArray();
                idx = idx + 1;
                c = c + 1;
            }
            while (c < ((Number)(n)).intValue()) {
                row = java.util.stream.DoubleStream.concat(java.util.Arrays.stream(row), java.util.stream.DoubleStream.of(0.0)).toArray();
                c = c + 1;
            }
            mat = appendObj(mat, row);
            r = r + 1;
        }
        printMat(mat);
    }

    static java.util.Map<String,Object> choleskyLower(java.util.Map<String,Object> a) {
        Object n = (Object)(((Object)a.get("order")));
        Object ae = (Object)(((Object)a.get("ele")));
        double[] le = new double[]{};
        int idx = 0;
        while (idx < String.valueOf(ae).length()) {
            le = java.util.stream.DoubleStream.concat(java.util.Arrays.stream(le), java.util.stream.DoubleStream.of(0.0)).toArray();
            idx = idx + 1;
        }
        int row = 1;
        int col = 1;
        int dr = 0;
        int dc = 0;
        int i = 0;
        while (i < String.valueOf(ae).length()) {
            Object e = ae[i];
            if (i < dr) {
                double d = (((Number)(e)).intValue() - le[i]) / le[dc];
le[i] = d;
                int ci = col;
                int cx = dc;
                int j = i + 1;
                while (j <= dr) {
                    cx = cx + ci;
                    ci = ci + 1;
le[j] = le[j] + d * le[cx];
                    j = j + 1;
                }
                col = col + 1;
                dc = dc + col;
            } else {
le[i] = sqrtApprox(((Number)(e)).intValue() - le[i]);
                row = row + 1;
                dr = dr + row;
                col = 1;
                dc = 0;
            }
            i = i + 1;
        }
        return new java.util.LinkedHashMap<String, Object>(java.util.Map.ofEntries(java.util.Map.entry("order", n), java.util.Map.entry("ele", le)));
    }

    static void demo(java.util.Map<String,Object> a) {
        System.out.println("A:");
        printSym(a);
        System.out.println("L:");
        java.util.Map<String,Object> l = choleskyLower(a);
        printLower(l);
    }
    public static void main(String[] args) {
        demo(makeSym(3, new double[]{25.0, 15.0, 18.0, -5.0, 0.0, 11.0}));
        demo(makeSym(4, new double[]{18.0, 22.0, 70.0, 54.0, 86.0, 174.0, 42.0, 62.0, 134.0, 106.0}));
    }

    static <T> T[] appendObj(T[] arr, T v) {
        T[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }
}

File: tests/rosetta/transpiler/Kotlin/cholesky-decomposition-1.kt
import java.math.BigInteger

var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Long {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        kotlin.math.abs(_nowSeed)
    } else {
        kotlin.math.abs(System.nanoTime())
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

fun sqrtApprox(x: Double): Double {
    var guess: Double = x
    var i: Int = 0
    while (i < 20) {
        guess = (guess + (x / guess)) / 2.0
        i = i + 1
    }
    return guess
}

fun makeSym(order: Int, elements: MutableList<Double>): MutableMap<String, Any?> {
    return mutableMapOf<String, Any?>("order" to (order), "ele" to (elements))
}

fun unpackSym(m: MutableMap<String, Any?>): MutableList<MutableList<Double>> {
    val n: Any? = (m)["order"] as Any?
    val ele: Any? = (m)["ele"] as Any?
    var mat: MutableList<MutableList<Double>> = mutableListOf<MutableList<Double>>()
    var idx: Int = 0
    var r: Int = 0
    while ((r).toInt() < (n as Number).toDouble()) {
        var row: MutableList<Double> = mutableListOf<Double>()
        var c: Int = 0
        while (c <= r) {
            row = run { val _tmp = row.toMutableList(); _tmp.add((ele[idx]).toDouble()); _tmp } as MutableList<Double>
            idx = idx + 1
            c = c + 1
        }
        while ((c).toInt() < (n as Number).toDouble()) {
            row = run { val _tmp = row.toMutableList(); _tmp.add(0.0); _tmp } as MutableList<Double>
            c = c + 1
        }
        mat = run { val _tmp = mat.toMutableList(); _tmp.add(row); _tmp } as MutableList<MutableList<Double>>
        r = r + 1
    }
    r = 0
    while ((r).toInt() < (n as Number).toDouble()) {
        var c: BigInteger = (r + 1).toBigInteger()
        while (c.compareTo(n.toBigInteger()) < 0) {
            mat[r][(c).toInt()] = mat[(c).toInt()][r]
            c = c.add(1.toBigInteger())
        }
        r = r + 1
    }
    return mat
}

fun printMat(m: MutableList<MutableList<Double>>): Unit {
    var i: Int = 0
    while (i < m.size) {
        var line: String = ""
        var j: Int = 0
        while (j < (m[i]).size) {
            line = line + (m[i][j]).toString()
            if (j < ((m[i]).size - 1)) {
                line = line + " "
            }
            j = j + 1
        }
        println(line)
        i = i + 1
    }
}

fun printSym(m: MutableMap<String, Any?>): Unit {
    printMat(unpackSym(m))
}

fun printLower(m: MutableMap<String, Any?>): Unit {
    val n: Any? = (m)["order"] as Any?
    val ele: Any? = (m)["ele"] as Any?
    var mat: MutableList<MutableList<Double>> = mutableListOf<MutableList<Double>>()
    var idx: Int = 0
    var r: Int = 0
    while ((r).toInt() < (n as Number).toDouble()) {
        var row: MutableList<Double> = mutableListOf<Double>()
        var c: Int = 0
        while (c <= r) {
            row = run { val _tmp = row.toMutableList(); _tmp.add((ele[idx]).toDouble()); _tmp } as MutableList<Double>
            idx = idx + 1
            c = c + 1
        }
        while ((c).toInt() < (n as Number).toDouble()) {
            row = run { val _tmp = row.toMutableList(); _tmp.add(0.0); _tmp } as MutableList<Double>
            c = c + 1
        }
        mat = run { val _tmp = mat.toMutableList(); _tmp.add(row); _tmp } as MutableList<MutableList<Double>>
        r = r + 1
    }
    printMat(mat)
}

fun choleskyLower(a: MutableMap<String, Any?>): MutableMap<String, Any?> {
    val n: Any? = (a)["order"] as Any?
    val ae: Any? = (a)["ele"] as Any?
    var le: MutableList<Double> = mutableListOf<Double>()
    var idx: Int = 0
    while (idx < ae.size) {
        le = run { val _tmp = le.toMutableList(); _tmp.add(0.0); _tmp } as MutableList<Double>
        idx = idx + 1
    }
    var row: Int = 1
    var col: Int = 1
    var dr: Int = 0
    var dc: Int = 0
    var i: Int = 0
    while (i < ae.size) {
        val e = ae[i]
        if (i < dr) {
            var d: Double = ((e as Number).toDouble() - le[i]) / le[dc]
            le[i] = d
            var ci: Int = col
            var cx: Int = dc
            var j: BigInteger = (i + 1).toBigInteger()
            while (j.compareTo(dr.toBigInteger()) <= 0) {
                cx = cx + ci
                ci = ci + 1
                le[(j).toInt()] = le[(j).toInt()] + (d * le[cx])
                j = j.add(1.toBigInteger())
            }
            col = col + 1
            dc = dc + col
        } else {
            le[i] = sqrtApprox((e as Number).toDouble() - le[i])
            row = row + 1
            dr = dr + row
            col = 1
            dc = 0
        }
        i = i + 1
    }
    return mutableMapOf<String, Any?>("order" to (n), "ele" to (le))
}

fun demo(a: MutableMap<String, Any?>): Unit {
    println("A:")
    printSym(a)
    println("L:")
    val l: MutableMap<String, Any?> = choleskyLower(a)
    printLower(l)
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        demo(makeSym(3, mutableListOf(25.0, 15.0, 18.0, 0.0 - 5.0, 0.0, 11.0)))
        demo(makeSym(4, mutableListOf(18.0, 22.0, 70.0, 54.0, 86.0, 174.0, 42.0, 62.0, 134.0, 106.0)))
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}
